flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResources.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.scheduler.declarative;
+
+import org.apache.flink.api.common.JobStatus;
+import org.apache.flink.runtime.executiongraph.ArchivedExecutionGraph;
+import org.apache.flink.runtime.executiongraph.ExecutionGraph;
+import org.apache.flink.runtime.jobmaster.slotpool.ResourceCounter;
+
+import org.slf4j.Logger;
+
+import javax.annotation.Nullable;
+
+import java.time.Duration;
+
+/**
+ * State which describes that the scheduler is waiting for resources in order to execute the job.
+ */
+class WaitingForResources implements State, ResourceConsumer {
+
+    private final Context context;
+
+    private final Logger logger;
+
+    private final ResourceCounter desiredResources;
+
+    WaitingForResources(Context context, Logger logger, ResourceCounter desiredResources) {
+        this.context = context;
+        this.logger = logger;
+        this.desiredResources = desiredResources;
+    }
+
+    @Override
+    public void onEnter() {
+        context.runIfState(this, this::resourceTimeout, Duration.ofSeconds(10L));
```

Comment:
If we merge this as is please file a ticket for making the timeout configurable and increasing the default to a higher value.

Reply:
Can we make it configurable right away? I think this would be good because it is also described like this in the latest state of the FLIP.